### DISCUSSION TITLE
How can I have a form collection with multiple fields and have the data stored in json with multiple fields per json object?

### DISCUSSION THREAD

**Discussion by jerrac** [symfony/symfony#47212] [Q&A] [answered]
Ok, I'm stuck. My google-fu has failed, and I could use another perspective.

(Sorry for the poor subject line... Really not sure how I could word it better...)

I want to let my users store notes on an entity. So I added a json field to the entity. If I add a collection type field to the form, I can use stimulus to let users add multiple text entries. That's all well and good, but I'd like to stick a datetime field into the form and json so they have an indication of when they made the note.

So my json data would look something like:
```json
[{

	"time": "2022-08-06 02:22:00",
	"text": "test"

}, {

	"time": "2022-08-06 02:34:00",
	"text": "test2"

}]
```
And the form would have a series of datetime input + textarea inputs. One for each note added.

I couldn't find a way to add extra fields to the `CollectionType`. I did figure out how to use `$builder->create()` with `FormType` to add the `DatetimeType` and `TextareaType` fields I wanted. But that falls apart when I start adding multiple notes. `FormType` doesn't have a prototype option, and the output form elements are named like we only have one note with just one time and one text field. So when my stimulus controller adds another form group (based on a manually created prototype), the two notes aren't sharing the same data structure like they would if I just had a `CollectionType` with a single input. Which, I think, is why my form doesn't submit properly when I have more than 1 note.

Here's a gist with my relevant code: https://gist.github.com/jerrac/7d5a5408baf15fa4a8d26f0c0b82a328

How should I go about doing this?

This is all in a new Symfony 6.1 project. I'm running PHP 8.1.

I've only been working with Symfony in my free time for month. So feel free to just point me at the proper place in the manual. I mean, I'd like to think I've scoured it for the answer, but I sure might have missed something... :)

Thanks in advance!

(Oh, one other question, how can I set the default value of a form field without overwriting existing data? Like setting my datetime field to "now".)

**Comment by Seb33300** [comment]
Hello @jerrac 

I think you are looking for Data Transformers:
https://symfony.com/doc/current/form/data_transformers.html

**Reply by jerrac** [reply]
I had read that page before. And I read it in detail yesterday, and again just now. I can almost see how Data Transformers could help, but it's just not clicking fully. 

Could you elaborate? I mean, my issue feels more like the form isn't being constructed properly, not that the data isn't being converted back and forth properly.

**Comment by jerrac** [comment, marked as answer]
Just figured it out on my own. :)

The solution was to create a `NotesType` class with the `DateTimeType` and `TextareaType` in it, then add a `CollectionType` element to my parent form with `entry_type` set to `NotesType::class`.

```php
$builder->add('notes', CollectionType::class, [
          'required' => false,
          'entry_type' => NotesType::class,
          'prototype' => true,
          'compound' => true,
          'allow_add' => true,
        ]);
```

After that I just had a few bugs in my Stimulus related code to iron out.
